apps/gov-website/pages/proposal/[pid].tsx

```diff
@@ -41,28 +70,142 @@ export const getStaticProps: GetStaticProps = async (context) => {
+				...proposal,
+				_id: proposal._id.toString(),
 			},
+			justification,
+			vetoThreshold,
 		},
 		revalidate: 600,
 	};
 };
 
 interface ProposalProps {
 	proposal: ProposalModel;
+	justification: string;
+	vetoThreshold: string;
 }
 
-const Proposal: NextPage<ProposalProps> = ({ proposal }) => {
-	const { proposalId } = proposal;
+const Proposal: NextPage<ProposalProps> = ({
+	proposal: initialProposal,
+	vetoThreshold,
+	justification,
+}) => {
+	const proposal = useProposal(initialProposal);
+	const { proposalId, call, status } = proposal;
+
+	const { open, openDialog, closeDialog } = useTransactionDialog();
+	const { onVote, onVeto, formState } = useProposalVoteForm(proposalId);
+
+	const onPass = useCallback(() => {
+		openDialog();
+		onVote(true);
+	}, [onVote, openDialog]);
+	const onReject = useCallback(() => {
+		openDialog();
+		if (status === "Deliberation") onVote(false);
+		if (status === "ReferendumDeliberation") onVeto();
+	}, [onVeto, onVote, openDialog, status]);
+
+	const onDialogDismiss = useCallback(() => {
+		closeDialog();
+	}, [closeDialog]);
+
+	useEffect(() => {
+		if (formState?.status === "Cancelled") closeDialog();
+	}, [closeDialog, formState?.status]);
+
+	const onDialogClose = useCallback(() => {
+		if (!formState?.status || formState?.status === "Cancelled") return;
+		onDialogDismiss();
+	}, [onDialogDismiss, formState?.status]);
 
 	return (
 		<Layout.PageWrapper>
 			<Header />
-			<Layout.PageContent>
+			<Layout.PageContent className="!max-w-4xl lg:!max-w-5xl">
 				<Layout.PageHeader>Proposal #{proposalId}</Layout.PageHeader>
+				<div className="grid grid-cols-1 gap-8 lg:grid-cols-3">
+					<div className="lg:col-start-3 lg:row-start-1">
+						<ProposalSidebar
+							proposal={proposal}
+							className="sticky top-12"
+							vetoThreshold={vetoThreshold}
+						/>
+					</div>
+					<div className="lg:col-span-2 lg:col-start-1 lg:row-start-1">
+						<ProposalBody.Justification justification={justification} />
+						<ProposalBody.Call call={call} />
+						<If
+							condition={
+								status === "Deliberation" || status === "ReferendumDeliberation"
+							}
+						>
+							<ProposalVoteForm
+								proposal={proposal}
+								onPass={onPass}
+								onReject={onReject}
+							/>
+						</If>
+					</div>
+				</div>
+				<ProposalVoteFormDialog
+					open={open}
+					formState={formState}
+					onClose={onDialogClose}
+					onDismiss={onDialogDismiss}
+				/>
 			</Layout.PageContent>
 		</Layout.PageWrapper>
 	);
 };
 
 export default Proposal;
+
+const useProposal = (initialProposal: ProposalModel) => {
+	const { api } = useCENNZApi();
+	const { proposalId } = initialProposal;
+	const [proposal, setProposal] = useState<ProposalModel>(initialProposal);
+
+	useEffect(() => {
+		if (!api) return;
+		let unsubscribeFn: VoidFn;
+		let lastBlockPolled: number;
+		api.rpc.chain
+			.subscribeFinalizedHeads(async (head) => {
+				const blockNumber = head.number.toNumber();
```

Comment:
could be worth extracting `subscribeFinalizedHeads` to a util that takes a callback?

Reply:
good call

Reply:
done